src/geom/stroke.rs

```diff
@@ -48,13 +87,219 @@ impl PartialStroke<Abs> {
 
 impl<T: Debug> Debug for PartialStroke<T> {
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        match (&self.paint, &self.thickness) {
-            (Smart::Custom(paint), Smart::Custom(thickness)) => {
-                write!(f, "{thickness:?} + {paint:?}")
+        let Self {
+            paint,
+            thickness,
+            line_cap,
+            line_join,
+            dash_pattern,
+            miter_limit,
+        } = &self;
+        if line_cap.is_auto()
+            && line_join.is_auto()
+            && dash_pattern.is_auto()
+            && miter_limit.is_auto()
+        {
+            match (&self.paint, &self.thickness) {
+                (Smart::Custom(paint), Smart::Custom(thickness)) => {
+                    write!(f, "{thickness:?} + {paint:?}")
+                }
+                (Smart::Custom(paint), Smart::Auto) => paint.fmt(f),
+                (Smart::Auto, Smart::Custom(thickness)) => thickness.fmt(f),
+                (Smart::Auto, Smart::Auto) => f.pad("<stroke>"),
+            }
+        } else {
+            write!(f, "(")?;
+            let mut sep = "";
+            if let Smart::Custom(paint) = &paint {
+                write!(f, "{}color: {:?}", sep, paint)?;
+                sep = ", ";
+            }
+            if let Smart::Custom(thickness) = &thickness {
+                write!(f, "{}thickness: {:?}", sep, thickness)?;
+                sep = ", ";
+            }
+            if let Smart::Custom(cap) = &line_cap {
+                write!(f, "{}cap: {:?}", sep, cap)?;
+                sep = ", ";
+            }
+            if let Smart::Custom(join) = &line_join {
+                write!(f, "{}join: {:?}", sep, join)?;
+                sep = ", ";
+            }
+            if let Smart::Custom(dash) = &dash_pattern {
+                write!(f, "{}dash: {:?}", sep, dash)?;
+                sep = ", ";
             }
-            (Smart::Custom(paint), Smart::Auto) => paint.fmt(f),
-            (Smart::Auto, Smart::Custom(thickness)) => thickness.fmt(f),
-            (Smart::Auto, Smart::Auto) => f.pad("<stroke>"),
+            if let Smart::Custom(miter_limit) = &miter_limit {
+                write!(f, "{}miter-limit: {:?}", sep, miter_limit)?;
+            }
+            write!(f, ")")?;
+            Ok(())
+        }
+    }
+}
+
+/// The line cap of a stroke
+#[derive(Clone, Eq, PartialEq, Hash)]
+pub enum LineCap {
+    Butt,
+    Round,
+    Square,
+}
+
+impl Debug for LineCap {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            LineCap::Butt => write!(f, "\"butt\""),
+            LineCap::Round => write!(f, "\"round\""),
+            LineCap::Square => write!(f, "\"square\""),
+        }
+    }
+}
+
+/// The line join of a stroke
+#[derive(Clone, Eq, PartialEq, Hash)]
+pub enum LineJoin {
+    Miter,
```

Comment:
Would it make sense to have the limit be part of this? Something like `Miter { limit: f64 }`. And on the Typst side it would be something like `#set line(stroke: (thickness: 1pt, join: (kind: "miter", limit: 4.0)))`. I'm unsure, it's getting a bit crazy nested at that point.

Reply:
I guess it could be an okay alternative, but I don't know if there is much point. It would introduce a little bit of extra complexity in both the docs and the implementation without so much gain, I don't think the `miter_limit` field in the stroke type is too much in the way.
 
A problem would also be how to set the `miter_limit` in a `set` rule without also setting the line join.